### DISCUSSION TITLE
Is there any chance to upgrade as3?

### DISCUSSION THREAD

**Discussion by badmfck** [airsdk/Adobe-Runtime-Support#897] [Ideas]
Is there any chance to upgrade as3 by adding generics, adding latest ecma scripts features?
Without improving and updating as3, there is no chance to get new developers, right now we are stuck in 2000.

**Comment by marchbold** [comment]
Would be great to see some language improvements! What ever happened to the as4 spec? Does that still exist?

**Reply by ajwfrost** [reply]
:-) I had wondered about this.. lots of options...

The AS4 spec was for a completely new virtual machine, so that would be a massive new effort. We had details of all of that from Adobe but as far as I know, there wasn't any prototype of the new VM that they had made, and we don't have a compiler for it....

For AS3, we do have source code for the AIR compiler, plus of course there's the Flex compiler which is open source. So it is theoretically possible for us to enhance the language and something that we could look at doing, although it would be fairly complex work.. I'd thought initially we could add some extra syntax support such as the 'lambda' function (so you can just declare an anonymous function inline as `(x, y) => x + y`).

Generics/templates may be possible as part of this capability is already present (i.e. the Vector class),

So yes, very interesting idea, we can have a look at this..

**Reply by ajwfrost** [reply]
One other - slightly far-out - thought that we actually had several years ago: what about if we integrated V8, such that we exposed all the AS3 APIs through JavaScript and you could then write your apps in JavaScript (or TypeScript perhaps for preference).

Wondering whether that would be popular or not. Personally I much prefer AS3, but it depends on what you're used to and what people are learning ..! Or, perhaps instead of having to embed the V8 engine into AIR, we could just create a compiler for TypeScript into ABC...

Currently of course, everything has to be compiled i.e. you get your ActionScript and turn it into ActionScript Byte Code within a SWF. There's no real "eval()" method, no compiler of ActionScript available at runtime. With JavaScript it's a whole different world and of course - well, it would have to be Apple WebKit - that would then mean you could actually dynamically execute scripts within AIR rather than having all the ahead-of-time compilation. (Although AOT does make the iOS start-up seem quite swift..!)

Anyway - would be interested in people's thoughts on this!

**Reply by IvakhnenkoDmytro** [reply]
Also would be great add some "syntax sugar" for null reference check like in C#. I mean instead of `if (_foo != null) _foo.bar()` use `_foo?.bar()`

Or even more - add something like Optional type in Swift or Nullable reference types in C#

**Reply by GeneralVimes** [reply]
> Or, perhaps instead of having to embed the V8 engine into AIR, we could just create a compiler for TypeScript into ABC...

If you add TypeScript/JavaScript without removing AS3 that'd be nice. Like, in Unity, it's possible to program on C# or on JS. If you remove AS3, leaving only JS, how will AIR stand out? If I wanted to write games on JS, I'd stick with Unity or Phaser.

I remember how I made a web version of my game on JS/Phaser, and then returned to AS3/AIR to build a PC/mobile version. What a relief it was to write on a good object-oriented compiled language again, which is AS3.

**Comment by ghost** [comment]


**Reply by ajwfrost** [reply]
Interesting!

The use of `Promise` there made me think for a minute whether there was a reason we couldn't add this sort of capability into the existing AIR runtime .. but someone's already done it:
https://github.com/CodeCatalyst/promise-as3

**Reply by marchbold** [reply]
That's really cool, would be great to have something like that as part of the as3 spec.

**Comment by 2jfw** [comment]
Adding new syntax/sugar etc could surely freshen up as3 and its acceptance and hopefully attract more (new) developers.
I would somehow think, that with a bunch of newly added syntax/functions/features and maybe a bit of performance optimizations for AVM2 you could call it **as4**, which would even have a bigger impact of perception.
Still loving the transparent and easy to understand syntax a lot and would love to see something new here!

**EDIT:** Also still thinking that **Harman** and **distriqt** should somehow "find" together to advance even more... 😉

**Reply by ajwfrost** [reply]
:-) somewhere I have the original specification for AS4 .. but that was going to be a whole new re-write of the virtual machine. I think incremental efforts are perhaps better...

**Reply by 2jfw** [reply]
I would pretty much think so :)
Just have a look at what other companies "do" in order to call their product "NEW" (🆕) - e.g. food companies changing their products ingredients slightly or car manufacturers doing face lifts... 

For as3: When there is a specific amount of impactful changes, it would be legit to call it **as4** ...

**Reply by GeneralVimes** [reply]
`<joke>`
Just add a function 
makeACoolGame() 
to ActionScript 4. This will for sure attract many fresh developers.
`</joke>`

**Comment by Michoko92** [comment]
Maybe I'm wrong, but isn't Haxe already another option for people wanting to use a more modern language?

**Reply by ghost** [reply]
Haxe is better than ActionScript, but isn't complete as JavaScript: still lacks destructuring patterns (don't know why) and asynchronous functions (available in HaxeLib as macro). It does have matching patterns though.

**Comment by usman48541** [comment]
AS3 was a great tool in its time, but it hasn’t kept up with modern programming features like generics or the latest ECMAScript standards. This makes it hard to attract new developers. Platforms like [Roblox ](https://robloxapkmod.com/)stay successful by regularly updating their tools and supporting modern programming methods, helping developers create exciting experiences. For AS3 to stay useful, it would need to do the same—add new features and meet what developers expect today. Without these updates, it’s hard for AS3 to compete with platforms that keep improving to match the needs of modern developers.
